apps/passport-client/components/modals/AccountImportModal.tsx

```diff
@@ -0,0 +1,187 @@
+import { useCallback, useEffect, useState } from "react";
+import styled from "styled-components";
+import { useFilePicker } from "use-file-picker";
+import { useDispatch, usePCDCollection } from "../../src/appHooks";
+import { Button } from "../core";
+
+type ImportState =
+  | {
+      valid: true;
+      imported: false;
+      pcds: PCD[];
+      folders: FolderMap;
+      conflicts: number;
+    }
+  | { valid: true; imported: true; added: number }
+  | { valid: false };
+
+type FolderMap = { [pcdId: string]: string };
+
+export function AccountImportModal() {
+  const { openFilePicker, filesContent } = useFilePicker({
+    accept: ".json",
+    multiple: false
+  });
+
+  const pcdCollection = usePCDCollection();
+  const dispatch = useDispatch();
+
+  const [importState, setImportState] = useState<ImportState | undefined>();
+
+  const importPCDs = useCallback(() => {
+    // Should never happen, but makes TypeScript happy that we checked for it
+    if (!importState.valid || importState.imported === true) return;
+
+    const addedPCDs: PCD[] = [];
+    for (const pcd of importState.pcds) {
+      if (!pcdCollection.hasPCDWithId(pcd.id)) {
+        try {
+          pcdCollection.add(pcd);
+          if (importState.folders[pcd.id]) {
+            pcdCollection.setFolder(pcd.id, importState.folders[pcd.id]);
+          }
+
+          addedPCDs.push(pcd);
+        } catch (e) {
+          console.log(e);
+        }
+      }
+
+      setImportState({ valid: true, imported: true, added: addedPCDs.length });
+    }
+  }, [importState, pcdCollection]);
+
+  useEffect(() => {
+    (async () => {
+      if (filesContent.length > 0) {
+        try {
+          const file = filesContent[0];
+          const data = JSON.parse(file.content);
+
+          if (data.pcds) {
+            const serializedPcdCollection = JSON.parse(data.pcds);
+            if (
+              serializedPcdCollection.pcds &&
+              serializedPcdCollection.folders
+            ) {
+              const pcds = await pcdCollection.deserializeAll(
+                serializedPcdCollection.pcds
+              );
+
+              const folders: { [pcdId: string]: string } =
+                serializedPcdCollection.folders;
+
+              let conflicts = 0;
+              for (const pcd of pcds) {
+                if (pcdCollection.hasPCDWithId(pcd.id)) {
+                  conflicts++;
+                }
+              }
+
+              setImportState({
+                valid: true,
+                imported: false,
+                pcds,
+                folders,
+                conflicts
+              });
+              return;
+            }
+          }
+
+          setImportState({ valid: false });
+        } catch (e) {
+          setImportState({ valid: false });
+        }
+      }
+    })();
+  }, [filesContent, pcdCollection]);
+
+  console.log(importState);
+
+  return (
+    <Container>
+      <Spacer h={24} />
+      {!importState && (
```

Comment:
A user-visible feature I'd consider adding to avoid more confusing issues:
Allow the user to import PCDs only in a single folder.  I think importing the Frogcrypto is going to be the majority use case, so providing that as an option would sidestep a lot of questions about proper handling of tickets and semaphore IDs.  Alternately, maybe we should just provide a separate "import frogs" button on the Frogcrypto page so users are encouraged to use that rather than bulk import.

Reply:
That would add a bit more UX complexity, but it does sound useful. I'll see how easy it might be to add this.